demos/meltdown_ac.cc

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Demonstrates Meltdown-AC - speculative fetching of unaligned data when
+ * alignment is enforced. This vulnerability seems to be AMD-specific. It should
+ * not work on Intel CPUs.
+ *
+ * We create an array of words, shift them by one byte to make them unaligned
+ * and then we copy the public and private data into respective unaligned
+ * arrays - one character is stored into one unaligned word.
+ * Afterwards we turn on the alignment enforcement and try to read the
+ * unaligned private array with that enforcement. That always leads to SIGBUS
+ * however the unaligned data are processed speculatively.
+ * It is necessary to have the AM (alignment mask) in CR0 register turned on,
+ * but on Linux it is a standard configuration. If the AM bit is off, the
+ * demonstration runs into the unreachable code and terminates with a failure.
+ **/
+
+#ifndef __linux__
+#  error Unsupported OS. Linux required.
+#endif
+
+#if !defined(__i386__) && !defined(__x86_64__)
+#  error Unsupported architecture. AMD required.
+#endif
+
+#include <array>
+#include <cstring>
+#include <fstream>
+#include <iostream>
+
+#include <signal.h>
+
+#include "cache_sidechannel.h"
+#include "instr.h"
+
+const char *public_data = "Hello, world!";
+const char *private_data = "It's a s3kr3t!!!";
+
+// Storage for the public data.
+size_t *public_array = new size_t[strlen(public_data) + 1];
```

Comment:
I think we're using `size_t` here because we need to make sure we're accessing something at least as large as the system's native word size. In that case I have a weak preference for `[u]intptr_t` as a "pointer-sized" integer type. Either way, let's leave a comment about why the type matters.

Reply:
Done.